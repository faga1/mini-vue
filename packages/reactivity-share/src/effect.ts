export let activeEffect: ReactiveEffect | undefined;
export const targetMap = new WeakMap<any, any>();
export class ReactiveEffect<T = any> {
  active = true;
  deps: Set<ReactiveEffect>[] = [];
  scheduler?: (fn: Function) => any = null;
  constructor(public fn: () => T) {}
  run() {
    activeEffect = this;
    return this.fn();
  }
  stop() {
    if (this.active === true) {
      this.deps.forEach(dep => {
        dep.delete(this);
      });
      this.active = false;
    }
  }
}
export function stop(runner: any) {
  return runner.effect.stop();
}
export function effect<T>(fn: () => T, options?: any) {
  const _effect = new ReactiveEffect<T>(fn);
  Object.assign(_effect, options);
  if (!options || !options.lazy) _effect.run();
  const runner = _effect.run.bind(_effect);
  runner.effect = activeEffect;
  return runner;
}
export function track(target: object, key: any) {
  if (activeEffect) {
    let depsMap = targetMap.get(target);
    if (!depsMap) {
      targetMap.set(target, (depsMap = new Map()));
    }
    let dep = depsMap.get(key);
    if (!dep) {
      depsMap.set(key, (dep = new Set()));
    }
    if (activeEffect.active) {
      dep.add(activeEffect);
    }
    activeEffect.deps.push(dep);
  }
}
export function trigger(target: object, key: any) {
  const depsMap = targetMap.get(target);
  if (!depsMap) return;
  const deps = depsMap.get(key);
  deps.forEach(effect => {
    if (effect.scheduler) {
      effect.scheduler(effect.run);
    } else {
      effect.run();
    }
  });
}
