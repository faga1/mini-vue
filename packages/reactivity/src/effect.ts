type EffectScheduler = (...args:any[]) => any
export interface ReactiveEffectOptions {
  lazy?: boolean;
  scheduler?: EffectScheduler;
  onStop?: () => void;
}

export interface ReactiveEffectRunner<T = any> {
  (): T;
  effect: ReactiveEffect;
}

export let activeEffect: ReactiveEffect | undefined;
export const targetMap = new WeakMap<any, any>();
export class ReactiveEffect<T = any> {
  active = true;
  deps: Set<ReactiveEffect>[] = [];
  onStop?: () => void;
  constructor(
    public fn: () => T,
    public scheduler:EffectScheduler | null = null
  ) {
  }
  run() {
    activeEffect = this;
    // cleanupEffect(this);
    return this.fn();
  }
  stop() {
    if (this.active) {
      cleanupEffect(this);
      this.active = false;
      if (this.onStop) {
        this.onStop();
      }
    }
  }
}

export function stop(runner: ReactiveEffectRunner) {
  runner.effect.stop();
}


export function effect<T>(fn: () => T, options?: ReactiveEffectOptions) {
  const _effect = new ReactiveEffect(fn);
  if (options) {
    Object.assign(_effect, options);
  }
  if (!options || !options.lazy) _effect.run();
  const runner = _effect.run.bind(_effect);
  runner.effect = _effect;
  return runner;
}

export function cleanupEffect(effect: ReactiveEffect){
  const { deps } = effect;
  if (deps.length) {
    for (let i = 0;i < deps.length; i++) {
      deps[i].delete(effect);
    }
    effect.deps = [];
  }
}
export function track(target: object, key:any) {
  if (activeEffect) {
    let depsMap = targetMap.get(target);
    if (!depsMap) {
      targetMap.set(target, (depsMap = new Map()));
    }
    let dep = depsMap.get(key);
    if (!dep) {
      depsMap.set(key, (dep = new Set()))
    }
    
    if (activeEffect.active) {
      dep.add(activeEffect);
    }
    activeEffect.deps.push(dep);
  }
}

export function trigger(target: object, key:any, oldValue?: unknown, newValue?: unknown) {     
  const depsMap = targetMap.get(target);
  if (!depsMap) return;
  const deps = depsMap.get(key);
  triggerEffects(deps);
}

function triggerEffects(effects: ReactiveEffect[]) {
  effects.forEach((effect) => triggerEffect(effect));
}

function triggerEffect(effect: ReactiveEffect) {
  if (effect.scheduler) {
    effect.scheduler()
  } else {
    effect.run()
  }
}