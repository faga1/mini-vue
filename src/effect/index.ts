// write your ts code

export let activeEffect = null;
export default function effect(fn){
    activeEffect = fn;
    fn();
}