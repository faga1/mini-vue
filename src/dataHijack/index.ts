import { activeEffect } from '../effect/index';
export let bucket:WeakMap<Object,bucketMap> = new WeakMap();
type bucketMap = Map<string|symbol,effectsSet>
type effectsSet = Set<Function>
export default function dataHijack(data){
    return new Proxy(data,{
        get(target,key){
            if(!activeEffect) return;
            let bucketMap = bucket.get(target)
            if(!bucketMap){
                bucketMap = new Map()
                bucket.set(target, bucketMap)
            }
            
            let effects = bucketMap.get(key);
            if(!effects){
                effects = new Set()
                bucketMap.set(key, effects)
            } 
            
            effects.add(activeEffect);
            
            return target[key]
        },
        set(target,key,newValue){
            target[key] = newValue;
            bucket.get(target)?.get(key).forEach(fn => fn())
            return newValue;
        }
    })
}
export function setBucket(obj){
    bucket.set(obj,new Map())
}