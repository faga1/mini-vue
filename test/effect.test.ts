import { assert, expect, it, describe } from 'vitest';
import effect from '../src/effect/index';
import dataHijack, { bucket, setBucket } from '../src/dataHijack';

const testFn = () => {
  console.log(res.a);
};

const obj1 = {
  a: 1,
  b: 2,
};
const res = dataHijack(obj1);
const relativeTest = obj => {
  effect(testFn);
  obj.a = 3;
  return bucket.get(obj1)?.get('a');
};
describe('relative', () => {
  it('obj', () => {
    expect(relativeTest(res)).toEqual(new Set([testFn]));
  });
});
