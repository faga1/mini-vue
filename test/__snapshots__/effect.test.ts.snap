// Vitest Snapshot v1

exports[`relative 1`] = `undefined`;
